src/transformStateWithClones.js

```diff
@@ -6,8 +6,39 @@
  *
  * @return {Object[]}
  */
-function transformStateWithClones(state, actions) {
-  // write code here
-}
+const transformStateWithClones = (state, actions) => {
+  const results = [Object.assign({}, state)];
+
+  for (const action of actions) {
+    const type = action.type;
+    let newState = Object.assign({}, results[results.length - 1]);
+
+    switch (type) {
+      case 'addProperties': {
+        newState = Object.assign(newState, action.extraData);
+        break;
+      }
+
+      case 'removeProperties': {
+        for (const key of action.keysToRemove) {
+          delete newState[key];
+        }
+        break;
+      }
+
+      case 'clear': {
+        newState = {};
+        break;
+      }
+
+      default: {
+        throw new Error('Something is wrong...');
```

Comment:
I think we should write there more specific error, for example 'Wrong action type' etc.

Reply:
Ok! Thank you!